src/Project/MvpSite/rendering/Middleware/CustomRenderingMiddleware.cs

```diff
@@ -0,0 +1,87 @@
+        {
+            _next = next;
+            _requestMapper = requestMapper;
+            _options = options.Value;
+            _graphQLClientFactory = graphQLClientFactory;
+            _graphQLRequestBuilder = graphQLRequestBuilder;
+            _layoutService = layoutService;
+            _configuration= configuration;
+        }              
+
+        public async Task Invoke(HttpContext httpContext, IViewComponentHelper viewComponentHelper, IHtmlHelper htmlHelper)
+        {
+            SitecoreLayoutRequest sitecoreLayoutRequest = _requestMapper.Map(httpContext.Request);
+
+            if (httpContext.GetSitecoreRenderingContext() == null)
+            {
+                SitecoreLayoutResponse response = await GetSitecoreLayoutResponse(httpContext).ConfigureAwait(continueOnCapturedContext: false);
+
+                //Check not found page and set status code - start
+                if (response.Request.Path() == _configuration.GetSection(MvpSiteSettings.Key).Get<MvpSiteSettings>().NotFoundPage) { httpContext.Response.StatusCode = (int)HttpStatusCode.NotFound; }
+                //Check not found page and set status code - end
+
+                SitecoreRenderingContext renderingContext = new()
+                {
+                    Response = response,
+                    RenderingHelpers = new RenderingHelpers(viewComponentHelper, htmlHelper)
+                };
+                httpContext.SetSitecoreRenderingContext(renderingContext);
+            }
+            else
+            {
+                httpContext.GetSitecoreRenderingContext().RenderingHelpers = new RenderingHelpers(viewComponentHelper, htmlHelper);
+            }
+
+            foreach (Action<HttpContext> postRenderingAction in (IEnumerable<Action<HttpContext>>)_options.PostRenderingActions)
+                postRenderingAction(httpContext);
+
+            httpContext.Items.Add((object)nameof(RenderingEngineMiddleware), (object)null);
+
+            await _next(httpContext).ConfigureAwait(continueOnCapturedContext: false);
+        }
+
+        private async Task<SitecoreLayoutResponse> GetSitecoreLayoutResponse(HttpContext httpContext)
+        {
+            //intercept for not found logic - start
+            CustomGraphQlLayoutServiceHandler customGraphQlLayoutServiceHandler = new(_configuration,_graphQLRequestBuilder, _graphQLClientFactory);
+            SitecoreLayoutRequest sitecoreLayoutRequest = _requestMapper.Map(httpContext.Request);
+            bool retVal = await customGraphQlLayoutServiceHandler.CheckLayoutExists(sitecoreLayoutRequest);
+            if (!retVal)  httpContext.Request.Path = _configuration.GetSection(MvpSiteSettings.Key).Get<MvpSiteSettings>().NotFoundPage;
+            //intercept for not found logic - end
+
+            sitecoreLayoutRequest = _requestMapper.Map(httpContext.Request);
+            return await _layoutService.Request(sitecoreLayoutRequest).ConfigureAwait(continueOnCapturedContext: false);
```

Comment:
The logic here isn't right, adding `awaitustomGraphQlLayoutServiceHandler.CheckLayoutExists(sitecoreLayoutRequest);` is causing an extra GQL request to be made to edge for every page request.

This needs to be changed to make the initial page request as normal, and then we should interrogate the response and only then make and extra request for the 404 page when required. That would remove the extra overhead being introduced here.